rpcs3/Emu/Memory/Memory.cpp

```diff
@@ -13,136 +13,94 @@ VirtualMemoryBlock* VirtualMemoryBlock::SetRange(const u32 start, const u32 size
-}
-
-u32 VirtualMemoryBlock::Map(u32 realaddr, u32 size)
-{
-	verify(HERE), (size);
-
-	for (u32 addr = m_range_start; addr <= m_range_start + m_range_size - 1 - GetReservedAmount() - size;)
-	{
-		bool is_good_addr = true;
-
-		// check if address is already mapped
-		for (u32 i = 0; i<m_mapped_memory.size(); ++i)
-		{
-			if ((addr >= m_mapped_memory[i].addr && addr < m_mapped_memory[i].addr + m_mapped_memory[i].size) ||
-				(m_mapped_memory[i].addr >= addr && m_mapped_memory[i].addr < addr + size))
-			{
-				is_good_addr = false;
-				addr = m_mapped_memory[i].addr + m_mapped_memory[i].size;
-				break;
-			}
-		}
-
-		if (!is_good_addr) continue;
-
-		m_mapped_memory.emplace_back(addr, realaddr, size);
-
-		return addr;
-	}
-
-	return 0;
+	return addr >= m_range_start && addr + size <= m_range_start + m_range_size - GetReservedAmount();
 }
 
 bool VirtualMemoryBlock::Map(u32 realaddr, u32 size, u32 addr)
 {
-	verify(HERE), (size);
-
-	if (!IsInMyRange(addr, size))
+	if (!size || !IsInMyRange(addr, size))
 	{
 		return false;
 	}
 
-	for (u32 i = 0; i<m_mapped_memory.size(); ++i)
-	{
-		if (addr >= m_mapped_memory[i].addr && addr + size - 1 <= m_mapped_memory[i].addr + m_mapped_memory[i].size - 1)
-		{
-			// it seems mapping another range inside a previous one is legit on ps3
-			// as long as it's coherent aliasing : offset from EA must match IO offset
-			// example game using this pattern : BCES01584 - the last of us
-			return (addr - m_mapped_memory[i].addr) == (realaddr - m_mapped_memory[i].realAddress);
-		}
-	}
-
 	m_mapped_memory.emplace_back(addr, realaddr, size);
 	return true;
 }
 
-bool VirtualMemoryBlock::UnmapRealAddress(u32 realaddr, u32& size)
+u32 VirtualMemoryBlock::UnmapRealAddress(u32 realaddr)
 {
 	for (u32 i = 0; i<m_mapped_memory.size(); ++i)
 	{
 		if (m_mapped_memory[i].realAddress == realaddr && IsInMyRange(m_mapped_memory[i].addr, m_mapped_memory[i].size))
 		{
-			size = m_mapped_memory[i].size;
+			const u32 size = m_mapped_memory[i].size;
 			m_mapped_memory.erase(m_mapped_memory.begin() + i);
-			return true;
+			return size;
 		}
 	}
 
-	return false;
+	return 0;
 }
 
-bool VirtualMemoryBlock::UnmapAddress(u32 addr, u32& size)
+u32 VirtualMemoryBlock::UnmapAddress(u32 addr)
 {
 	for (u32 i = 0; i<m_mapped_memory.size(); ++i)
 	{
 		if (m_mapped_memory[i].addr == addr && IsInMyRange(m_mapped_memory[i].addr, m_mapped_memory[i].size))
 		{
-			size = m_mapped_memory[i].size;
+			const u32 size = m_mapped_memory[i].size;
 			m_mapped_memory.erase(m_mapped_memory.begin() + i);
-			return true;
+			return size;
 		}
 	}
 
-	return false;
+	return 0;
 }
 
-bool VirtualMemoryBlock::Read32(const u32 addr, u32* value)
+s64 VirtualMemoryBlock::Read32(const u32 addr)
 {
-	u32 realAddr;
-	if (!getRealAddr(addr, realAddr))
-		return false;
-	*value = vm::read32(realAddr);
-	return true;
+	if (const u32 realAddr = getRealAddr(addr))
+	{
+		return static_cast<u64>(vm::read32(realAddr));
```

Comment:
'Read32' but your returning s64?

Reply:
its not a mistake, im using the higher 32 bits of the return value to tell if the read was successful.

Reply:
the cast is to make sure the compiler dosn't sign extend the value.